cmd/kubeadm/app/cmd/token.go

```diff
@@ -255,8 +263,19 @@ func RunGenerateToken(out io.Writer) error {
 	return nil
 }
 
+type outputItem struct {
```

Comment:
I dislike this type in its current form:

- It is basically a copy of [BoostrapToken](https://github.com/kubernetes/kubernetes/blob/3c9c7fca64808fb305f5b2b59dea8c4eab6aa4cc/cmd/kubeadm/app/apis/kubeadm/v1beta1/types.go#L227) with a couple of additional fields attached to it.

- It's not versioned, which means that we don't guarantee to users, that kubeadm 1.15 and 1.16 will deliver compatible or even predictable outputs.

Reply:
@rosti Thank you for review!

yes, having it not versioned is not good, I agree. However, using only BootstrapToken would make the solution half-baked in my opinion. The main purpose of this is to provide a convenient way to get *all* required info to run kubeadm join. I believe this will be appreciated by the projects that have to parse output of 'kubeadm init' to get this info.

I'm open to suggestions how to make it better. Should we rename outputItem type to something more recognisable and make it versioned?

Reply:
how about adding a struct that wraps BoostrapToken with the extra 2 fields?

Reply:
I tried that. It shows token as a struct in a JSON/YAML output. I didn't like it.
However, I'd say it's better than nothing.

Reply:
I agree with @rosti that we should aim at having this versioned as well.
However, I'm a little bit reluctant in linking this to the current config API. I have to think a little bit more about this

Reply:
I am thinking on something along the lines of:
```
type BootstrapTokenList struct {
  metav1.TypeMeta `json:",inline"`
  PublicKeyPins        []string `json:"publicKeyPins"`
  ControlPlaneHostPort string   `json:"controlPlaneHostPort"`
  tokens []kubeadmapiv1beta1.BootstrapToken `json:"tokens"`
}
```

However, I am too reluctant to add this to the config API. It's not an API, it's set in stone structured output format.
@bart0sh @fabriziopandini @neolit123 WDYT?

Reply:
i would omit the TypeMeta for BootstrapTokenList

Reply:
@rosti I like the idea. Not sure I support adding metav1.TypeMeta, but the rest looks good to me. Should we wrap this into high-level structure to make JSON/YAML output to contain  'BootstrapTokenList'? This way the output would clearly say what type it is. WDYT, guys?

Reply:
Even if we use `kubeadmapiv1beta1.BootstrapToken` for the bootstrap token slice here, we won't be able to state to the user, that these are v1beta1 or v1beta2 `BootstrapTokens`.
On the other hand, we can safely drop TypeMeta, but we need to document the exact format of this structure.

Reply:
I tried using this type:
```
type BootstrapTokenList struct {
	PublicKeyPins        []string                     //`yaml:"Publickeypins"`
	ControlPlaneHostPort string                       //`yaml:"ControlPlaneHostport"`
	Tokens               []*kubeadmapi.BootstrapToken //`yaml:"Tokens"`
}
```
and wrapped it into unnamed struct and the result looks good to me:
```
root@kind-control-plane:/# kubeadm token list --output json
{
  "TokenList": {
    "PublicKeyPins": [
      "sha256:5f50bc74bd59d2ad0063e7aa639a0c1161dbde0101cb499b33e9881f37b60d77"
    ],
    "ControlPlaneHostPort": "172.17.0.2:6443",
    "Tokens": [
      {
        "Token": "abcdef.0123456789abcdef",
        "Description": "",
        "TTL": null,
        "Expires": "2019-03-27T22:16:04Z",
        "Usages": [
          "authentication",
          "signing"
        ],
        "Groups": [
          "system:bootstrappers:kubeadm:default-node-token"
        ]
      },
      {
        "Token": "d4clx6.jlw4j9zdubqlul9c",
        "Description": "",
        "TTL": null,
        "Expires": "2019-03-27T22:26:44Z",
        "Usages": [
          "authentication",
          "signing"
        ],
        "Groups": [
          "system:bootstrappers:kubeadm:default-node-token"
        ]
      }
    ]
  }
}
```

However,  YAML result is lower-cased:
```
root@kind-control-plane:/# kubeadm token list --output yaml
tokenlist:
  publickeypins:
  - sha256:5f50bc74bd59d2ad0063e7aa639a0c1161dbde0101cb499b33e9881f37b60d77
  controlplanehostport: 172.17.0.2:6443
  tokens:
  - token:
      id: abcdef
      secret: 0123456789abcdef
    description: ""
    ttl: null
    expires: "2019-03-27T22:16:04Z"
    usages:
    - authentication
    - signing
    groups:
    - system:bootstrappers:kubeadm:default-node-token
  - token:
      id: d4clx6
      secret: jlw4j9zdubqlul9c
    description: ""
    ttl: null
    expires: "2019-03-27T22:26:44Z"
    usages:
    - authentication
    - signing
    groups:
    - system:bootstrappers:kubeadm:default-node-token
```

And the only way to make it look the same as JSON is to use `yaml:"MixedCaseFieldName"` or `json:"lowercasefieldname"` in kubeadmapi.BootstrapToken definition. Is it ok to do? Any other ideas how to make them look similar? should we care at all?

Reply:
ideally the marshaled output should be `camelCase` for both json and yaml: as in the `json` tags of the kubeadm api: https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/apis/kubeadm/v1beta1/types.go

Reply:
@neolit123, @rosti You seem to think about this in a two different ways. For @neolit123(and for me too) it's just an improved output format that's easy to parse. For @rosti it's kinda API, which should be versioned to ensure its stability. We should agree on what we want it to be to proceed further I guess.

Reply:
i mentioned earlier that `BootstrapTokenList` should not be treated as an API object.
but if we marshal it, i'm pretty sure we want consistency in the fields between JSON and YAML to be `camelCase`.

Reply:
done. please review.